### PR TITLE
The UI Testing Booster Pack

### PR DESCRIPTION
**Pull request by gdostie**
https://gdostie.github.io/source.coveo.com/

### CODE REVIEW

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+- It displays a count of zero at the start and an "edit" button
+- When clicking on "edit" button
+  - It replaces the "edit" button by "plus", "minus", and "save" buttons
+  - It increments the count by one when clicking on the "plus" button
+  - It decrements the count by one when clicking on the "minus" button
+  - It replaces the "plus", "minus", and "save" buttons by "edit" button when clicking on the "save" button
+
+#### Step 2: Write down all test cases titles
+
+Spending time thinking about the test cases titles might be one of the most overlooked step in the testing world. However, doing so is crutial to establish a strong understanding of how the test cases should be split. Basically, the idea is to write the most detailed, succinct, and non-ambiguous description of each behaviours you have identified on step 1. Having them all written down before even thinking about the code needed to make the test work will help you understand how to split your tests. The scope of each of them will become much clearer in your head when comes the time to write their underlying code.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:59]: Typo: doing so is crutial => doing so is crucial

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+For our `Counter` example, I have basically already have done all the work on Step 1, but here's how it would look in the code:
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:61]: This sounds a bit weird :sweat: Maybe just:

I have basically already have done => I already have done all the work on Step 1

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+Talking about confidence, _"The more your tests resemble the way your software is used, the more confidence they can give you."_. It is the guiding principle of [Testing Library](https://testing-library.com/), the library we are using to test our UIs. In other words it means that our tests must only do stuff that a real user of the interface can do. Doing otherwise leads to tests that rely on implementation details. Tests that rely on implementation details will generates false negatives (behaviour did not change and test broke) and false positives (behaviour changed and test passed). Having false positives and false negatives reduce the confidence we have in our tests and generate additional maintenance costs. Less value plus higher cost equals lower return on investment and quite franckly a drop in happiness for everyone involved. Lets see how that principle applies to our counter example.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:95]: Nitpick: In other words, it means (missing comma)

Another nitpick: replace "do stuff" by "do actions" :sweat_smile: 

franckly => frankly

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+```tsx
+describe('when clicking on "edit" button', () => {
+  it('replaces the "edit" button by "plus", "minus", and "save" buttons', () => {
+    // TODO
+  });
+});
+```
+
+We can copy-paste the setup portion from the previous test. We already know how to get the edit button since we did this in the previous test, so from there its not much more work to trigger a click on the same element using [`user-event`](https://testing-library.com/docs/ecosystem-user-event/). At this point we reach unknown territories, but we can apply the same strategy as before involving the testing playground.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:158]: so from there, it's not much (comma and "it's")

At this point, we reach (comma)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+  );
+
+  const editButton = screen.getByRole("button", { name: /edit/i });
+  userEvent.click(editButton);
+
+  screen.logTestingPlaygroundURL();
+});
+```
+
+Now the screen looks like [this](https://testing-playground.com/#markup=DwEwlgbgfMAOUBUAWBTABAYwPYFcB2ALmmAM4BcaADMAPTzABGOBBWeUA1LUy2zD63YBabs0H8xfAMoBDCClG92tcNCA), and we end up with a test that is similar to the first one but verifies a different behaviour.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:176]: Now, the screen (comma)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+  ).not.toBeInTheDocument();
+  expect(screen.getByRole("button", { name: "+" })).toBeInTheDocument();
+  expect(screen.getByRole("button", { name: "-" })).toBeInTheDocument();
+  expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();
+});
+```
+
+The thinking process to wrap up the other test cases is exactly the same. We can extract the repeating setup portion into a `beforeEach` block if we want, but it is not mandatory and not always a good practice.
+
+> Unlike what we've been accustomed to do in source code, copy-pasting blocks of code in each tests is totally fine for the sake of readability and test independance. [Here](https://mtlynch.io/good-developers-bad-tests/) is a great read on that subject that I highly recommend.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:205]: I think I'd replace "Here" in the link with a more descriptive text, like the name of the Article :thinking:

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+#### Step 5: Doubt your tests
+
+Once we're all done writting the test cases we have envisionned at the start, we could stop there and call it a day. Truthfully that would be totally acceptable. In an context where one must rush out a feature, just testing the happy path with the most straightforward assertions is a whole lot better than having no tests at all. However, when developping a critical component that needs to stand the test of time, going just one step further will future proof your work against refactors and regressions. To be honnest, at the beginning this step can mean added time to delivery, but with experience we start writing better code and better tests right from the start.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:209]: To be honest, at the beginning, this step (add a comma, remove extra n in honnest)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+  );
+
+  userEvent.click(screen.getByRole("button", { name: /edit/i }));
+  userEvent.click(screen.getByRole("button", { name: "+" }));
+
+  expect(store.getState().counter.count).toBe(1);
+});
+```
+
+The test passes just fine and does validate that the count has been incremented by one. Now, what if someone refactors the component to remove redux altogether while retaining all the functionalities? There wouldn't be any `store` object to expect from anymore. The test would fail while the behaviour hasn't changed creating a false negative. The problem here is that our test relies on an implementation detail: redux. Would our user see if the count has changed by looking into the state object himself? Of course not, the user only looks at the screen.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:229]: > The problem here is that our test relies on an implementation detail: redux.

Reading this, this doesn't seem right. The Redux update can be, in some way, one of the "output" of the component!

What if another component relies on this state, and then we remove it? The component doesn't output any state, so other components could break because of that.

So I'd suggest not saying that Redux is strictly an implementation detail. *Yes* it can be an detail, but it can also be an output, just use proper judgement :smile:

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+  render(<MyComponent />);
+  const emailInput = screen.getByRole('textbox', {name: 'email'});
+  userEvent.type(emailInput, 'invalid email address');
+  expect(screen.getByText('Invalid email')).toBeInTheDocument();
+});
+```
+
+Don't get me wrong, testing the `isValidEmail` util separately is still a very good idea. It gives you confidence that all the edge cases are covered and that wherever the util is used, it will work as it is expected to. In the components that use the util we could only test the most common branches. For example one test for a valid email and one test for an invalid email. After all we only care about if the validation message is displayed when the right conditions occur, so that would be enough.
+
+If you like that idea of mocking fewer things as possible, it is even possible to stop mocking the API client entirely. I won't go into details about this one since we have yet to try this in our team. If that interests you, [here](https://kentcdodds.com/blog/stop-mocking-fetch)'s a great article about it.
```
Comment (francoislg) [inline _posts/2021-06-21-UI-testing-booster-pack.md:362]: nitpick: I prefer to replace `here` with the actual article name

**Review by francoislg — APPROVED** [review]
Amazing! I love this! Great article and makes a good case for `react-testing-library` :smile:

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 7 above)
Comment (amoreauCoveo) [reply _posts/2021-06-21-UI-testing-booster-pack.md:209]: writing (extra t)
envisioned (extra n)
In a context (extra n)
developing (extra p)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+```ts
+expect(screen.getByText("The count is: 1")).toBeInTheDocument();
+```
+
+The above assertion is a lot stronger since it validates what the user really sees. We could refactor the component to use any library and the test would still pass, as long as the functionnality is preserved. Yes the `<Provider>` would probably have to be removed, but it wouldn't hurt the test's logic to leave it there as well.
+
+Pushing the doubt further: what if a new intern (I always imagine an intern in those scenarios 😄) breaks the `Counter`'s logic so that when pressing on the "+" button, it doesn't increment more than once anymore. So from 0 it goes to 1, then stays at 1 when clicking on "+" again. Our current test wouldn't pick up on that breakage of functionality, leading to a false positive.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:237]: I would prefer "on that breaking of functionality"

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+
+  userEvent.click(plusButton);
+  expect(screen.getByText("The count is: 1")).toBeInTheDocument();
+
+  userEvent.click(plusButton);
+  expect(screen.getByText("The count is: 2")).toBeInTheDocument();
+});
+```
+
+We can go on with questions like this until we are satisfied by what we have given the confidence level that we are aiming for. Sometimes we will discover new behaviours that were overlooked at the beginning. Feel free to add more tests cases or rethink how you split the previous test cases.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:257]: test cases (extra s)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+
+## Bonus tips for the road
+
+If you have made it this far congratulations! You are ready to roll on the UI testing highway to no regression land. This section regroups tips and tricks in random order that I noticed helped me write stronger tests.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:261]: If you have made it this far, congratulations! (added comma)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+
+It can be tempting to reuse the same util in your tests assertions, thinking that if someone changes the string it won't break our test.
+
+```jsx
+expect(
+  screen.getByRole("button", { name: translate("Save") })
+).toBeInTheDocument();
+```
+
+However by doing so, you are more subject false positives. Imagine someone (probably an intern again) mistakenly changes the "Save" translation for "Cancel". I know this is a bit extreme, but it is just to illustrate my point. In that case the test would still pass, because `translate` would return the same wrong string both in the component and the test. The UI would then save when clicking on "Cancel" which you would guess is not what we want!
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:279]: you are more subject to false positives (missing "to")

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+
+> A visible string that changes, is a break in functionality.
+
+```jsx
+expect(screen.getByRole("button", { name: /'save'/i })).toBeInTheDocument();
+```
+
+Hardcoding the "save" string directly into the test protects us from that potential breakage. The same principle applies for URLs, numbers, dates and other visible values you could think of.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:287]: numbers, dates, and other . . . (missing Oxford comma)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+it("should do X", () => {
+  // ...
+});
+
+it("should not do Y", () => {
+  // ...
+});
+```
+
+Repeating the same word in each test case is not a problem per say, but it does not bring any value. On the other hand, getting rid of it has advantages. In addition to having one less word to write (more succinct), it removes ambibuity and forces you to find a clearer phrasing. The UI _should_ not do _X_, it _does_ _X_. If it doesn't do X, it is not acceptable, the behaviour is broken and the test has to fail.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:303]: ambiguity (misspelled)

the behaviour is broken, and the test . . .

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+
+#### Mock as fewer things as possible
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:305]: "as little things" sounds better I think

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+});
+```
+
+It's better than no tests at all, but at least two things make this a weak test:
+1. It doesn't validate that `isValidEmail` is called with the input value (potential false positive)
+2. If the component is refactored to use a different validator function (for example an external lib), we will get a false negative
+
+Sure we could fix the first problem by asserting the mock is called with the input value we type in, but when using the mock strategy, it doesn't force you to think about adding these additional assertions so they become really easy to forget. You also would still have problem number 2.
+
+The solution to both problems is to make your test ressemble the way a user would use the interface and drop the mock.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:349]: resemble (extra s)

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 9 above)
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:362]: Same comment as title

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+});
+```
+
+Don't get me wrong, testing the `isValidEmail` util separately is still a very good idea. It gives you confidence that all the edge cases are covered and that wherever the util is used, it will work as it is expected to. In the components that use the util we could only test the most common branches. For example one test for a valid email and one test for an invalid email. After all we only care about if the validation message is displayed when the right conditions occur, so that would be enough.
+
+If you like that idea of mocking fewer things as possible, it is even possible to stop mocking the API client entirely. I won't go into details about this one since we have yet to try this in our team. If that interests you, [here](https://kentcdodds.com/blog/stop-mocking-fetch)'s a great article about it.
+
+## Closing remarks
+
+Writing tests for UIs is like doing fine arts. Tests come in many colours, lines can get pretty abstract and every artist has his own style. This article's purpose was to help you conquer your blank page syndrome rather than telling you which colour is the best.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:366]: lines can get pretty abstract, and every artist has their own style (gender neutral + missing comma)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+
+If there is one sentence you need to remember from this post, it would be [Testing Library](https://testing-library.com/)'s guiding principle: _"The more your tests resemble the way your software is used, the more confidence they can give you"_.
+
+
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:370]: Could you try adding a nice HR thing? 😄 
Like this:

`Do you like this type of challenge? Do you think we're missing something in our tests? [Join the Coveo Team](https://www.coveo.com/en/company/careers/open-positions) and help us improve the search experience!`

**Review by amoreauCoveo — APPROVED** [review]
:gg:
a few comments mostly about spelling
but good job :)

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+    <Provider store={store}>
+      <Counter />
+    </Provider>
+  );
+
+  screen.logTestingPlaygroundURL();
+});
+```
+
+Now the test passes and gives us a nice [playground link](https://testing-playground.com/#markup=DwEwlgbgfMAOUBUAWBTABAYwPYFcB2ALmmAM4BcaADMAPTzABGOBBWeUAouAbUy2zBrhoQ) thanks to the [screen.logTestingPlaygroundURL](https://testing-library.com/docs/queries/about/#screenlogtestingplaygroundurl) method. This utility makes it very easy to visualize what is being rendered and gives us good understanding of what can be asserted from there. By leveraging the testing playground, following the [best practices](https://testing-library.com/docs/queries/about#priority) and avoiding the [common mistakes](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library), writing good assertions becomes pretty straightforward.
```
Comment (amoreauCoveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:128]: I just realized that you wrote "visualize" here with a z, but "minimise" with an s earlier. You can't have both.
You would either need to write "visualize" and "minize" to use Canadian spelling, or "visualise" and "minimise" to use British spelling.

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 8 above)
Comment (GermainBergeron) [reply _posts/2021-06-21-UI-testing-booster-pack.md:229]: I actually think it should be another test if we want to test that the component modifies the state, with a clearer name. This way it's easier to understand and make sure that the UI still displays the correct thing in one test and make sure we modify the state in another

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+```jsx
+import {isValidEmail} from './validators';
+
+const MyComponent = () => {
+  // other stuff
+  const validationMessage = isValidEmail(inputValue) ? '' : 'Invalid email';
+
+  return (
+    // other stuff
+    {validationMessage ? <span className='invalid'>{validationMessage}</span> : null}
```
Comment (GermainBergeron) [inline _posts/2021-06-21-UI-testing-booster-pack.md:324]:
```suggestion
  const errorMessage = isValidEmail(inputValue) ? '' : 'Invalid email';

  return (
    // other stuff
    {errorMessage ? <span className='invalid'>{errorMessage}</span> : null}
```

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 22]
```diff
@@ -0,0 +1,370 @@
...
+Sure we could fix the first problem by asserting the mock is called with the input value we type in, but when using the mock strategy, it doesn't force you to think about adding these additional assertions so they become really easy to forget. You also would still have problem number 2.
+
+The solution to both problems is to make your test ressemble the way a user would use the interface and drop the mock.
+
+```jsx
+it("displays an invalid message when the email is invalid", () => {
+  render(<MyComponent />);
+  const emailInput = screen.getByRole('textbox', {name: 'email'});
+  userEvent.type(emailInput, 'invalid email address');
+  expect(screen.getByText('Invalid email')).toBeInTheDocument();
```
Comment (GermainBergeron) [inline _posts/2021-06-21-UI-testing-booster-pack.md:356]: This is a bit confusing since the user types `invalid email` and we make sure `Invalid email` is in the document
```suggestion
  userEvent.type(emailInput, 'not an email');
  expect(screen.getByText('Invalid email')).toBeInTheDocument();
```

**Review by GermainBergeron — APPROVED** [review]
Really good blog post! 🎉 🥇

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 8 above)
Comment (francoislg) [reply _posts/2021-06-21-UI-testing-booster-pack.md:229]: :point_up: totally agree with that! I just want to be sure that we don't suggest to **not** test redux states :smile:

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 23]
```diff
@@ -0,0 +1,370 @@
...
+
+tags: [UI, tests, react, testing library, frontend]
+
+author:
+  name: Gaël Dostie
+  bio: Software Developer, Administration Console UI team
+  image: gdostie.jpeg
+---
+
+You feel like you do a decent job at creating new user intefaces or modifying existing ones, but writing tests to cover your changes takes you more time than to write the actual source code. You have experience writing tests for clear cut units of code like functions or classes, but it seems that testing UI is just too different. Having lost all intuition about what needs to be tested, what needs to be mocked, how to split your test cases or even how to write a simple assertion, you feel lost and unproductive. If you can relate to any of the above, this article is for you.
```
[the remark this answers is not in the document]
Comment (kdang-coveo) [reply _posts/2021-06-21-UI-testing-booster-pack.md:14]:
```suggestion
You feel like you do a decent job at creating new user interfaces or modifying existing ones, but writing tests to cover your changes takes you more time than to write the actual source code. 

You have experience writing tests for clear cut units of code like functions or classes, but it seems that testing UI is just too different. Having lost all intuition about what needs to be tested, what needs to be mocked, how to split your test cases or even how to write a simple assertion, you feel lost and unproductive. If you can relate to any of the above, this article is for you.
```

Optional spacing here - I personally like breaking up into smaller paragraphs (kind of like how medium posts are):

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 24]
```diff
@@ -0,0 +1,370 @@
...
+it('displays a count of zero at the start and an "edit" button', () => {
+  const store = createStore(combineReducers({ counter: CounterReducer }));
+  render(
+    <Provider store={store}>
+      <Counter />
+    </Provider>
+  );
+
+  expect(screen.getByText("The count is: 0")).toBeInTheDocument();
+  expect(screen.getByRole("button", { name: /edit/i })).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:140]: I think we could, although I'm not sure I fully understand the difference between the two. I know `toBeVisible` takes into account the css if any is loaded (`display:none` amongst other rules). Is that it?

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 25]
```diff
@@ -0,0 +1,370 @@
...
+
+<!-- more -->
+
+## Why do we write tests
+
+Let's start with _why_. If you think about it, when we create or change something visual, we surely look at it and interact with it manualy while changing its source code until we are satisfied with the result. So why do we bother creating additional scripted tests that don't have eyes to tell us if our interfaces look good or bad?
+
+A simple answer to that question is that scripted tests (unit, functional, end-to-end, all of them, we'll get to that part later) provide us with a certain level of confidence that the UI (still) works as intended.
+
+> In a sense, scripted tests are a written garantee that the UI under test respects a given set of behaviours.
```
Comment (kdang-coveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:26]:
```suggestion
> In a sense, scripted tests are a written guarantee that the UI being tested under certain circumstances are being respected.
```

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+     title="redux-counter"
+     sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
+   ></iframe>
+
+#### Step 1: Identify the behaviours
+
+The first step to find out what needs to be tested is to understand how the component behaves. The best way to do that is usually to interact with the component in a demo environment. To start out, look at the initial state, that is what is first being rendered. Once you know what the initial state is, you need to find every way you can make this initial state change into a derived state. From each of those derived state, if the component behaves exactly the same way compared to the initial state, you have found all the behaviours. On the contrary, if from a derived state the component behaves differently, you have identified new behaviours that will also need to be tested.
+
+After looking at what's initialy rendered and clicking on the available buttons, we can assume that there are 5-6 behaviours depending on how we choose to split them. Don't forget that the initial state is in itself a behaviour.
```
Comment (kdang-coveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:48]:
```suggestion
After looking at what's initially rendered and clicking on the available buttons, we can assume that there are 5-6 behaviours depending on how we choose to split them. Don't forget that the initial state is in itself a behaviour.
```

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+- It displays a count of zero at the start and an "edit" button
+- When clicking on "edit" button
+  - It replaces the "edit" button by "plus", "minus", and "save" buttons
+  - It increments the count by one when clicking on the "plus" button
+  - It decrements the count by one when clicking on the "minus" button
+  - It replaces the "plus", "minus", and "save" buttons by "edit" button when clicking on the "save" button
+
+#### Step 2: Write down all test cases titles
+
+Spending time thinking about the test cases titles might be one of the most overlooked step in the testing world. However, doing so is crucial to establish a strong understanding of how the test cases should be split. Basically, the idea is to write the most detailed, succinct, and non-ambiguous description of each behaviour you have identified on step 1. Having them all written down before even thinking about the code needed to make the test work will help you understand how to split your tests. The scope of each of them will become much clearer in your head when comes the time to write their underlying code.
```
Comment (kdang-coveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:59]:
```suggestion
Spending time thinking about the test cases titles might be one of the most overlooked step in the testing world. However, doing so is crucial to establish a strong understanding of how the test cases should be split. Basically, the idea is to write the most detailed, succinct, and non-ambiguous description of each behaviour you have identified on Step 1. Having them all written down before even thinking about the code needed to make the test work will help you understand how to split your tests. The scope of each of them will become much clearer in your head when comes the time to write their underlying code.
```

Capitalizing for consistency since you had it capitalized below ;D

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+    });
+
+    it('replaces the "plus", "minus", and "save" buttons by "edit" button when clicking on the "save" button', () => {
+      // TODO
+    });
+  });
+});
+```
+
+When dealing with derived states, like the one we have after clicking on the _Edit_ button, there are more than one valid solution to split the test cases. We could only use `it` blocks with longer titles or leverage `describe` blocks to shorten similar titles. In the end, it all comes down to personal or team preferences. I personally try to minimize the usage of `describe` blocks, since it reduces readability of test case titles when you have a lot of them. You need to go back up the file to read the `describe` bloc to understand the full test.
```
Comment (kdang-coveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:89]:
```suggestion
When dealing with derived states, like the one we have after clicking on the _Edit_ button, there are more than one valid solution to split the test cases. We could only use `it` blocks with longer titles or leverage `describe` blocks to shorten similar titles. 

In the end, it all comes down to personal or team preferences. I personally try to minimize the usage of `describe` blocks, since it reduces readability of test case titles when you have a lot of them. You need to go back up the file to read the `describe` bloc to understand the full test.
```

Optional: You can separate this into separate paragraph since it can act as a mini conclusion here

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 29]
```diff
@@ -0,0 +1,371 @@
...
+
+Talking about confidence, _"The more your tests resemble the way your software is used, the more confidence they can give you."_. It is the guiding principle of [Testing Library](https://testing-library.com/), the library we are using to test our UIs. In other words, it means that our tests must only perform actions that a real user of the interface can do. Doing otherwise leads to tests that rely on implementation details. Tests that rely on implementation details will generate false negatives (behaviour did not change and test broke) and false positives (behaviour changed and test passed). Having false positives and false negatives reduce the confidence we have in our tests and generates additional maintenance costs. Less value plus higher cost equals lower return on investment and quite frankly a drop in happiness for everyone involved. Let's see how that principle applies to our counter example.
```
Comment (kdang-coveo) [inline _posts/2021-06-21-UI-testing-booster-pack.md:95]:
```suggestion
Talking about confidence,

> "The more your tests resemble the way your software is used, the more confidence they can give you."

It is the guiding principle of [Testing Library](https://testing-library.com/), the library we are using to test our UIs. In other words, it means that our tests must only perform actions that a real user of the interface can do. Doing otherwise leads to tests that rely on implementation details. Tests that rely on implementation details will generate false negatives (behaviour did not change and test broke) and false positives (behaviour changed and test passed). Having false positives and false negatives reduce the confidence we have in our tests and generates additional maintenance costs. Less value plus higher cost equals lower return on investment and quite frankly a drop in happiness for everyone involved. Let's see how that principle applies to our counter example.
```

Suggestion to use a quote block here

**Comment by kdang-coveo** [comment]
Really awesome write-up 🙏 

This would definitely be something I needed to read when I first started here at Coveo! 💯

**Review by kdang-coveo — APPROVED** [review]

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md]: The first 3 sentences are like an enumeration so I makes more sense to me to keep them close. I moved `If you can relate to any of the above, this article is for you.` to the second paragraph though. Thanks!

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 30]
```diff
@@ -0,0 +1,370 @@
...
+
+Running the test above gives us the following error.
+
+> Could not find "store" in the context of "Connect(Counter)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(Counter) in connect options.
+
+Without even looking at the component's implementation, it is pretty obvious from the error message that we are missing something: a redux store provider. Looking at the component's implementation as little as possible forces us to not rely on implementation details, which is really the key to get better tests as explained earlier.
+
+```tsx
+it('displays a count of zero at the start and an "edit" button', () => {
+  const store = createStore(combineReducers({ counter: CounterReducer }));
```
[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:117]: I could start by using the "application" store and then in the `Doubt` section I would explain how using the real "application" store leads to tests that aren't independent, because it leads to tests that reuse a store that has been potentially modified by previous tests. Then I could reswitch to redefining the store in each test. Thoughts?

[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:117]: I think I will create a section for it in the bonus tips part. To explain de duality between implementation details and test independence.

[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:125]: Done

File: _posts/2021-06-21-UI-testing-booster-pack.md [hunk 31]
```diff
@@ -0,0 +1,371 @@
...
+
+If you like that idea of mocking as little things as possible, it is even possible to stop mocking the API client entirely. I won't go into details about this one since we have yet to try this in our team. If that interests you, [Stop mocking fetch](https://kentcdodds.com/blog/stop-mocking-fetch) is a great article on the subject.
+
+## Closing remarks
+
+Writing tests for UIs is like doing fine arts. Tests come in many colours, lines can get pretty abstract, and every artist has their own style. This article's purpose was to help you conquer your blank page syndrome rather than telling you which colour is the best.
+
+If there is one sentence you need to remember from this post, it would be [Testing Library](https://testing-library.com/)'s guiding principle: _"The more your tests resemble the way your software is used, the more confidence they can give you"_.
```
[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:415]: Since I already did a block quote for the exact same sentence earlier, I just felt it looked better like this 🤷

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 25 above)
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:26]: You suggestion has a different meaning I think 🤔

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md]: I ended up switching them all to `toBeVisible`

File: _posts/2021-06-21-UI-testing-booster-pack.md (hunk 8 above)
Comment (gdostie) [reply _posts/2021-06-21-UI-testing-booster-pack.md:237]: Added a paragraph to explain exactly that

**Review by francoislg — APPROVED** [review]
:confetti_ball:
